### PR TITLE
Fixed an issue where port number was not added to host header.

### LINKED ISSUE #3721: S3 Transfer Manager host/endpoint mismatch
### Describe the bug

I'm using localstack and notice that changing the `endpointOverride` causes failures with transfer with message:
```
Cannot create meta s3 request; host header value a.b.c.d does not match endpoint override a.b.c.d:4566
```

### Expected Behavior

S3 Transfer Manager should work.

### Current Behavior

Library:
```
s3-transfer-manager v2.19.25
aws-crt v0.21.2
```

Error:
```
Exception in thread "main" java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.)
    at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:65)
...
    at software.amazon.awssdk.transfer.s3.S3TransferManager.upload(S3TransferManager.java:447)
    at org.example.TransferBug.main(TransferBug.java:31)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.)
    at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
    at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:47)
...
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncExecutionFailureExceptionReportingStage.execute(AsyncExecutionFailureExceptionReportingStage.java:41)
    ... 18 more
Caused by: software.amazon.awssdk.crt.CrtRuntimeException: S3Client.aws_s3_client_make_meta_request: creating aws_s3_meta_request failed (aws_last_error: AWS_ERROR_INVALID_ARGUMENT(34), An invalid argument was passed to a function.) AWS_ERROR_INVALID_ARGUMENT(34)
    at software.amazon.awssdk.crt.s3.S3Client.s3ClientMakeMetaRequest(Native Method)
    at software.amazon.awssdk.crt.s3.S3Client.makeMetaRequest(S3Client.java:79)
...
    at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryingExecutor.attemptExecute(AsyncRetryableStage.java:144)
    ... 25 more
```

I've tried various endpoint, and got the same result:
```
[ERROR] [2023-01-27T03:21:10Z] [000000016b70b000] [S3Client] - id=0x121f0f010 Cannot create meta s3 request; host header value 127.0.0.1 does not match endpoint override 127.0.0.1:4566
```

```
[ERROR] [2023-01-27T03:21:27Z] [000000016d0e7000] [S3Client] - id=0x11ee09010 Cannot create meta s3 request; host header value fake-test-bucket.localhost does not match endpoint override fake-test-bucket.localhost:4566
```

```
[ERROR] [2023-01-27T03:21:49Z] [000000016b68f000] [S3Client] - id=0x142e0caa0 Cannot create meta s3 request; host header value fake-test-bucket.s3.localhost.localstack.cloud does not match endpoint override fake-test-bucket.s3.localhost.localstack.cloud:4566
```

### Reproduction Steps

Running the code on localstack throws the error.

```
package org.example;

import java.net.URI;

import software.amazon.awssdk.core.async.AsyncRequestBody;
import software.amazon.awssdk.crt.Log;
import software.amazon.awssdk.services.s3.S3AsyncClient;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.transfer.s3.S3TransferManager;

public class TransferBug {
    public static void main(String... args) {
        Log.initLoggingToFile(Log.LogLevel.Error, "file-output.txt");

        var endpoint = "http://s3.localhost.localstack.cloud:4566";
        var bucket = "fake-test-bucket";
        var key = "test.txt";

        var client = S3AsyncClient.crtBuilder()
                .endpointOverride(URI.create(endpoint))
                .build();

        try (var s3TransferManager = S3TransferManager.builder()
                .s3Client(client)
                .build()
        ) {
            var putRequest = PutObjectRequest.builder()
                    .bucket(bucket).key(key)
                    .build();

            var upload = s3TransferManager.upload(b -> b.putObjectRequest(putRequest)
                    .requestBody(AsyncRequestBody.fromString("test")));

            upload.completionFuture().join();
        }
    }
}
```

docker-compose.yaml
```
services:
  # https://github.com/localstack/localstack/blob/v1.3.1/docker-compose.yml
  localstack:
    image: localstack/localstack:1.3.1
    ports:
      - "127.0.0.1:4566:4566"            # LocalStack Gateway
      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
    environment:
      - DEBUG=${DEBUG-}
      - LAMBDA_EXECUTOR=${LAMBDA_EXECUTOR-}
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-./volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
```

### Possible Solution

I downgraded `aws-crt` to `v0.21.1`, and it works for some endpoints.

* `s3.localhost.localstack.cloud` => works
* `127.0.0.1` => works
* `localhost` => fails, likely due to missing `pathStyleAccessEnabled`

The issue appears to be caused by https://github.com/awslabs/aws-crt-java/pull/581, and not sure if it is intentional.

### Additional Information/Context

_No response_

### AWS Java SDK version used

2.19.25

### JDK version used

openjdk version "17.0.4" 2022-07-19 LTS; OpenJDK Runtime Environment Corretto-17.0.4.8.1 (build 17.0.4+8-LTS)

### Operating System and version

MacOS 12.6 (Monterey)

### PR DESCRIPTION
**Pull request by zoewangg**
## Motivation and Context
Fixed an issue where port number was not added to host header.
Closes #3721 #3682 

## Modifications
Add port to the host number

## Testing
Added unit tests
Running integ tests and stability tests
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java [hunk 1]
```diff
@@ -27,15 +27,16 @@
...
 import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
 import software.amazon.awssdk.services.s3.S3AsyncClientBuilder;
 import software.amazon.awssdk.services.s3.S3Client;
 import software.amazon.awssdk.services.s3.S3ClientBuilder;
+import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
 import software.amazon.awssdk.services.s3.model.GetBucketPolicyResponse;
 
 public class GetBucketPolicyFunctionalTest {
-    private static final URI HTTP_LOCALHOST_URI = URI.create("http://localhost:8080/");
```
Comment (cenedhryn) [inline services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java:38]: Do you have the bandwidth to convert these two classes to JUnit5? It'd be helpful for our migration. [BlobUnmarshallingTest.java](https://github.com/aws/aws-sdk-java-v2/blob/master/test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/protocolrestjson/BlobUnmarshallingTest.java) is an example with both parameterized tests and Wiremock.

**Review by cenedhryn — COMMENTED** [review]
Looks good!

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java (hunk 1 above)
Comment (zoewangg) [reply services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java:38]: Unfortunately no, I agree and I'd normally do it for new tests. Do you think this is a blocker?

**Review by cenedhryn — APPROVED** [review]

File: services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java (hunk 1 above)
Comment (cenedhryn) [reply services/s3/src/test/java/software/amazon/awssdk/services/s3/functionaltests/GetBucketPolicyFunctionalTest.java:38]: No, that's fine.

**Pull request merged**